# BREAST CANCER DETECTION USNG RANDOM FOREST TREE

In [2]:
# Breast cancer detection using RANDOM FOREST TREE 

# import library
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier 
from sklearn.metrics import confusion_matrix, accuracy_score
import numpy as np

# load data
data = pd.read_csv("E:/Ducat/Machine Learning/Breast_cancer.csv")

# Split data into features (X) and target (y)
X, y = data.iloc[:, :-1], data.iloc[:, -1]

# Train-test split
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

# Train Random Forest model
random_forest = RandomForestClassifier(random_state=0)  # Initialize the RandomForestClassifier
random_forest.fit(x_train, y_train)  # Train the model

# Make predictions
y_pred = random_forest.predict(x_test)  # Predict the classes for the test set

# Evaluate the model
result = confusion_matrix(y_test, y_pred)  # Get the confusion matrix
print('Confusion Matrix:')
print(result)
print('Accuracy Score:', accuracy_score(y_test, y_pred))  # Print the accuracy score

# Predict for a single sample
k = np.array([[15, 10, 5, 100, 0.1]])

# Convert k to a DataFrame with the same columns as X
k_df = pd.DataFrame(k, columns=X.columns)  # Ensure k has the same feature names as X

# Predict the class for k using the DataFrame
pred = random_forest.predict(k_df)  # Predict the class for the input data
print("Prediction for Breast Cancer:", pred)  # Print the prediction for the sample


Confusion Matrix:
[[ 56   7]
 [  3 105]]
Accuracy Score: 0.9415204678362573
Prediction for Breast Cancer: [1]


In [41]:
# import all library
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier 
from sklearn.metrics import confusion_matrix, accuracy_score
import numpy as np

In [8]:
# load and read the dataset
df = pd.read_csv("E:/Ducat/Machine Learning/Breast_cancer.csv")
df

mean_radius  mean_texture  mean_perimeter  mean_area  mean_smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     diagnosis  
0            0  
1            0  
2            0  
3            0  
4            0  
..         ...  
564          0  
565          0  
566          0  
567          0  
568          1  

[569 rows x 6 columns]

In [16]:
# Select feature columns
X = data.iloc[:, :-1]
X

mean_radius  mean_texture  mean_perimeter  mean_area  mean_smoothness
0          17.99         10.38          122.80     1001.0          0.11840
1          20.57         17.77          132.90     1326.0          0.08474
2          19.69         21.25          130.00     1203.0          0.10960
3          11.42         20.38           77.58      386.1          0.14250
4          20.29         14.34          135.10     1297.0          0.10030
..           ...           ...             ...        ...              ...
564        21.56         22.39          142.00     1479.0          0.11100
565        20.13         28.25          131.20     1261.0          0.09780
566        16.60         28.08          108.30      858.1          0.08455
567        20.60         29.33          140.10     1265.0          0.11780
568         7.76         24.54           47.92      181.0          0.05263

[569 rows x 5 columns]

In [18]:
# Select target columns
y = data.iloc[:, -1]
y

0      0
1      0
2      0
3      0
4      0
      ..
564    0
565    0
566    0
567    0
568    1
Name: diagnosis, Length: 569, dtype: int64

In [20]:
# Train-test split
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

In [22]:
# Train Random Forest model
random_forest = RandomForestClassifier(random_state=0)  # Initialize the RandomForestClassifier
random_forest.fit(x_train, y_train)  # Train the model

RandomForestClassifier(random_state=0)

In [24]:
# Make predictions
y_pred = random_forest.predict(x_test)  # Predict the classes for the test set

In [26]:
# Evaluate the model
result = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(result)
print('Accuracy Score:', accuracy_score(y_test, y_pred))

Confusion Matrix:
[[ 56   7]
 [  3 105]]
Accuracy Score: 0.9415204678362573


In [28]:
# Predict for a single sample (adjust k to match the number of features)
# Example sample data with the same number of features as the training set
k = np.array([[15, 10, 5, 100, 0.1]])  # Example input data for prediction with 5 features

In [30]:
# Convert k to a DataFrame with the same columns as X
k_df = pd.DataFrame(k, columns=X.columns)  # Ensure k has the same feature names as X

In [32]:
# Predict the class for k using the DataFrame
pred = random_forest.predict(k_df)  # Predict the class for the input data
print("Prediction for Breast Cancer:", pred)  # Print the prediction for the sample

Prediction for Breast Cancer: [1]


# Breast Cancer Detection using Random Forest

This project uses the **Random Forest** algorithm to detect breast cancer. Below are the steps involved in the project:

### Steps Involved:
1. **Importing Libraries**:  
   Essential libraries such as `pandas`, `sklearn`, and `numpy` are imported for data handling, machine learning, and evaluation.

2. **Loading the Dataset**:  
   The dataset is loaded into a Pandas DataFrame from a CSV file. It contains features and a target column indicating cancer diagnosis (e.g., `0` for benign, `1` for malignant).

3. **Data Splitting**:  
   - The data is split into features (`X`) and the target (`y`).  
   - Further split into training and testing sets using `train_test_split`, with 30% of data allocated for testing.

4. **Training the Random Forest Model**:  
   - A `RandomForestClassifier` is initialized and trained on the training set (`x_train`, `y_train`).  
   - The trained model is used to make predictions on the test set (`x_test`).

5. **Model Evaluation**:  
   - The performance is evaluated using a **confusion matrix** and the **accuracy score**.  
   - The confusion matrix provides details of correct and incorrect predictions, while the accuracy score represents overall model accuracy.

6. **Single Sample Prediction**:  
   - A sample input is created as a NumPy array and converted to a DataFrame with the same feature names as the dataset.  
   - The trained model predicts the class (benign or malignant) for this sample input.

---

### Output Summary:
- **Confusion Matrix**:
 This matrix indicates strong model performance with minimal misclassifications.

- **Accuracy Score**:  
`94.15%` – A high level of accuracy in predicting breast cancer.

- **Prediction for Sample Input**:  
`[1]` – Indicates malignant cancer for the given sample input.

